File: backend/src/prices.py
"""Cached Star Citizen ore prices, read from the published UEX price feed.

The feed (a small prices.json on GitHub Pages) is produced hourly by the prices
workflow from UEX Corp's public data. This cache fetches it on a timer so we
never hit UEX per scan; if a fetch fails we keep the last good values.

Price data: UEX Corp (https://uexcorp.space).
"""

import asyncio
import json
import logging
import time
import urllib.request
from typing import Dict, Optional

logger = logging.getLogger(__name__)


class PriceCache:
    """Holds the latest ore prices fetched from the feed URL."""

    def __init__(self, url: str, refresh_seconds: int = 3600):
        self.url = url
        self.refresh_seconds = refresh_seconds
        self.prices: Dict[str, dict] = {}   # ore_id -> {name, sell, buy}
        self.updated_at: Optional[int] = None
        self.last_fetch: float = 0.0

    def _apply(self, data: dict) -> None:
        """Apply a parsed feed payload to the cache (separated for testing)."""
        self.prices = data.get("prices", {}) or {}
        self.updated_at = data.get("updated_at")
        self.last_fetch = time.time()

    def _fetch(self) -> None:
        req = urllib.request.Request(self.url, headers={"User-Agent": "sc-ore-scanner"})
        with urllib.request.urlopen(req, timeout=15) as resp:
            self._apply(json.load(resp))
        logger.info(f"Loaded {len(self.prices)} ore prices from {self.url}")

    async def refresh(self) -> None:
        """Refresh the cache; on failure keep the previously cached prices."""
        try:
            await asyncio.to_thread(self._fetch)
        except Exception as e:
            logger.warning(f"Price refresh failed (keeping cached prices): {e}")

    def sell_price(self, ore_id: str) -> Optional[int]:
        entry = self.prices.get(ore_id)
        return entry.get("sell") if entry else None

    def value_of(self, ore_id: str, quantity: int) -> Optional[int]:
        sell = self.sell_price(ore_id)
        return sell * quantity if sell else None

    def summary(self) -> dict:
        return {
            "url": self.url,
            "updated_at": self.updated_at,
            "count": len(self.prices),
            "prices": self.prices,
        }

File: core/src/prices.rs
//! Ore price cache — reads the published UEX feed (a small prices.json on GitHub
//! Pages) over HTTP and caches it. Mirrors the Python PriceCache: fetch on a
//! timer (the caller decides cadence), never per scan; keep last good values on
//! failure. Price data: UEX Corp (https://uexcorp.space).

use std::collections::HashMap;
use std::io::Read;
use std::time::Duration;

use serde::Deserialize;

pub const DEFAULT_FEED_URL: &str = "https://hunter-36.github.io/sc-ore-scanner/prices.json";

/// Default connect/read timeout for the feed fetch. It runs on the scan thread
/// (startup + hourly), so an unbounded request would stall the overlay; bound it.
const DEFAULT_TIMEOUT: Duration = Duration::from_secs(10);

/// Hard cap on the response body. The real feed is a few KB; this guards against
/// a runaway or hostile response exhausting memory. An oversized body is read up
/// to the cap, then fails to parse — so the cache is preserved (see `refresh`).
const MAX_FEED_BYTES: u64 = 4 * 1024 * 1024;

#[derive(Deserialize, Clone, Default)]
pub struct PriceEntry {
    #[serde(default)]
    pub name: String,
    #[serde(default)]
    pub sell: i64,
    #[serde(default)]
    pub buy: i64,
}

#[derive(Deserialize, Default)]
struct Feed {
    #[serde(default)]
    prices: HashMap<String, PriceEntry>,
    #[serde(default)]
    updated_at: Option<i64>,
}

pub struct PriceCache {
    url: String,
    prices: HashMap<String, PriceEntry>,
    pub updated_at: Option<i64>,
    timeout: Duration,
}

impl PriceCache {
    pub fn new(url: impl Into<String>) -> Self {
        Self {
            url: url.into(),
            prices: HashMap::new(),
            updated_at: None,
            timeout: DEFAULT_TIMEOUT,
        }
    }

    /// Override the network timeout (primarily for tests).
    pub fn with_timeout(mut self, timeout: Duration) -> Self {
        self.timeout = timeout;
        self
    }

    /// Fetch the feed and replace the cache. On error — including a connect/read
    /// timeout or an over-cap response — the previous cache is kept untouched.
    pub fn refresh(&mut self) -> anyhow::Result<()> {
        let agent = ureq::AgentBuilder::new()
            .timeout_connect(self.timeout)
            .timeout_read(self.timeout)
            .build();
        let resp = agent.get(&self.url).call()?;

        // Cap the body: read at most MAX_FEED_BYTES. An oversized feed is
        // truncated here and then fails JSON parsing, so we never accept it and
        // the cache (assigned only after a successful parse) survives.
        let mut body = String::new();
        resp.into_reader()
            .take(MAX_FEED_BYTES)
            .read_to_string(&mut body)?;
        let feed: Feed = serde_json::from_str(&body)?;
        self.prices = feed.prices;
        self.updated_at = feed.updated_at;
        Ok(())
    }

    /// Sell price per SCU for an ore id, if known and > 0.
    pub fn sell_price(&self, ore_id: &str) -> Option<i64> {
        self.prices.get(ore_id).map(|p| p.sell).filter(|&s| s > 0)
    }

    pub fn len(&self) -> usize {
        self.prices.len()
    }

    pub fn is_empty(&self) -> bool {
        self.prices.is_empty()
    }
}
